pkg/reconciler/contour/contour.go

```diff
@@ -69,65 +68,31 @@ type Reconciler struct {
-var _ controller.Reconciler = (*Reconciler)(nil)
+var _ ingressreconciler.Interface = (*Reconciler)(nil)
 
-// Reconcile implements controller.Reconciler
-func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+func (r *Reconciler) ReconcileKind(ctx context.Context, ingress *v1alpha1.Ingress) reconciler.Event {
 	logger := logging.FromContext(ctx)
-	ctx = r.configStore.ToContext(ctx)
-
-	// Convert the namespace/name string into a distinct namespace and name
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		logger.Errorf("Invalid resource key: %q.", key)
-		return nil
-	}
-
-	// If our controller has configuration state, we'd "freeze" it and
-	// attach the frozen configuration to the context.
-	//    ctx = r.configStore.ToContext(ctx)
 
 	// Get the resource with this namespace/name.
-	original, err := r.lister.Ingresses(namespace).Get(name)
-	if apierrs.IsNotFound(err) {
-		// The resource may no longer exist, in which case we stop processing.
-		logger.Errorf("Resource %q no longer exists.", key)
-		return nil
-	} else if err != nil {
-		return err
-	} else if original.Annotations != nil {
-		class := original.Annotations[networking.IngressClassAnnotationKey]
+	if ingress.Annotations != nil {
+		class := ingress.Annotations[networking.IngressClassAnnotationKey]
 		if class != ContourIngressClassName {
-			logger.Debugf("Resource %q is not our class.", key)
+			logger.Debugf("Resource %q is not our class.", ingress.Name)
 			return nil
 		}
 	}
-	// Don't modify the informers copy.
-	resource := original.DeepCopy()
 
-	// Reconcile this copy of the resource and then write back any status
-	// updates regardless of whether the reconciliation errored out.
-	reconcileErr := r.reconcile(ctx, resource)
-	if equality.Semantic.DeepEqual(original.Status, resource.Status) {
-		// If we didn't change anything then don't call updateStatus.
-		// This is important because the copy we loaded from the informer's
-		// cache may be stale and we don't want to overwrite a prior update
-		// to status with this stale state.
-	} else if _, err = r.updateStatus(resource); err != nil {
-		logger.Warnw("Failed to update resource status.", zap.Error(err))
-		r.recorder.Eventf(resource, corev1.EventTypeWarning, "UpdateFailed",
-			"Failed to update status for %q: %v", resource.Name, err)
-		return err
-	}
+	reconcileErr := r.reconcile(ctx, ingress)
+	// reconcileErr := r.reconcileIngress(ctx, ingress)
 	if reconcileErr != nil {
-		r.recorder.Event(resource, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+		logger.Errorf("Failed to reconcile Ingress %s", ingress.Name, reconcileErr)
+		//ingress.Status.MarkIngressNotReady(notReconciledReason, notReconciledMessage)
+		return reconcileErr
 	}
-	return reconcileErr
+	return reconciler.NewEvent(corev1.EventTypeNormal, "IngressTypeReconciled", "IngressType reconciled: \"%s/%s\"", ingress.Namespace, ingress.Name)
```

Comment:
I wasn't doing this across serving because it feels noisy, and I'm not sure how to use it well yet.